python/oneseismic/scan/tests/test_scan.py

```diff
@@ -40,3 +49,432 @@ def test_regular_intervals(inlines, crosslines):
+            from_big_endian(header[int(segyio.su.xline)])
+        ])
+        cdps = np.array([
+            header[int(segyio.su.cdpx)],
+            header[int(segyio.su.cdpy)],
+            1
+        ])
+        estimated_linenos = utm_to_lineno.dot(cdps)
+
+        np.testing.assert_allclose(estimated_linenos, linenos, atol=1e-9)
+
+
+@given(
+    integers(min_value = 2, max_value = 15),
+    integers(min_value = 2, max_value = 15),
+    one_of(
+        floats(min_value = -10, max_value = -1),
+        floats(min_value = 1, max_value = 10)
+    ),
+    one_of(
+        floats(min_value = -10, max_value = -1),
+        floats(min_value = 1, max_value = 10)
+    ),
+    floats(min_value = -math.pi, max_value = math.pi),
+    one_of(
+        floats(min_value = -math.pi / 2, max_value = -1e-3),
+        floats(min_value = 1e-3, max_value = math.pi / 2)
+    ),
+    floats(min_value = -10, max_value = 10),
+    floats(min_value = -10, max_value = 10)
+)
+def test_geo_scanner_lines_not_perpendicular(
+        inlines,
+        crosslines,
+        ilinc,
+        xlinc,
+        rotation,
+        shear,
+        offsetx,
+        offsety
+):
+    """
+    shear
+        A rotation added to the crossline direction to make it non-pependicular
+        to the inline direction
+
+    See test_geo_scanner for a description of the other parameters
+    """
+    ilincx = ilinc * math.cos(rotation)
+    ilincy = ilinc * math.sin(rotation)
+    xlincx = - xlinc * math.sin(rotation + shear)
+    xlincy = xlinc * math.cos(rotation + shear)
+
+    headers = [
+        {
+            int(segyio.su.iline): big_endian(i),
+            int(segyio.su.xline): big_endian(x),
+            int(segyio.su.cdpx):
+                i * ilincx + x * xlincx + offsetx,
+            int(segyio.su.cdpy):
+                i * ilincy + x * xlincy + offsety,
+            int(segyio.su.ns): 0,
+            int(segyio.su.dt): 0,
+            int(segyio.su.scalco): 1,
+        }
+        for i in range(inlines + 1)
+        for x in range(crosslines + 1)
+    ]
+
+    seg = GeoScanner(
+        il_word=int(segyio.su.iline),
+        xl_word=int(segyio.su.xline),
+        endian=sys.byteorder,
+    )
+
+    for header in headers:
+        seg.scan_trace_header(header)
+
+    result = seg.report()
+
+    # Scanner should return an empty dict when assumption of perpendicular lines
+    # doesn't hold
+    assert result == {}
+
+
+@given(
+    integers(min_value = 4, max_value = 15),
+    integers(min_value = 4, max_value = 15),
+    one_of(
+        floats(min_value = -10, max_value = -1),
+        floats(min_value = 1, max_value = 10)
+    ),
+    one_of(
+        floats(min_value = -10, max_value = -1),
+        floats(min_value = 1, max_value = 10)
+    ),
+    floats(min_value = -math.pi, max_value = math.pi),
+    floats(min_value = -10, max_value = 10),
+    floats(min_value = -10, max_value = 10)
+)
+def test_geo_scanner_out_of_order_and_missing_lines(
```

Comment:
It is because it some times select 3 points on a line. I will change it to select min(inlines, crosslines) + 1 so that we are guaranteed that at least 1 point is not on the same line.